### PR TITLE
ManyBabies AtHome - internationalization (merged changes from dev)

### PR DESCRIPTION
**Pull request by rhodricusack**

### CODE REVIEW

**Comment by kimberscott** [comment]
Ok, did a first round of proper review! I made some minor changes (in the PR back to this branch) to make sure the study-detail template renders & headers are translated along with remaining text. 

This added a few new {% trans ... %} blocks so will require additional entries in the translation too - I'm not familiar with makemessages / compilemessages (...or Japanese) so haven't touched the .po file.

0) This is a dumb question, but what steps do you take to try out the translations displaying locally? When I set Firefox or Chrome to use Japanese as its first/default language, I don't get any translations when looking at localhost:8000. I've generated the django.po file in locale/ja/LC_MESSAGES using `django-admin compilemessages`. 

1) It looks like all of the participant-facing text is translated, but not all of the researcher-facing text (some templates under accounts/, plus the templates in `exp/` and `studies/`. I think this makes sense for getting started - there's much less participant-facing text, it's more stable, and it's critical to translate for parents to participate (whereas researchers who can figure out our docs in English can figure out the interface in English). Just wanted to confirm this was intentional. If so, @Datamance - any opinions on having *some* translation in the researcher-facing text vs. holding off on translating the researcher-facing text at all?

2) There's a bit of text defined in `web/views.py` (success messages like "You've successfully logged out" and error messages lines 292-293 if someone goes to a study that's no longer running) and `web/templates/web/study-detail.html` (the "X days/months/years until DATE" text constructed in Javascript on line 53)

3) @Datamance - preference on checking in the compiled translation files (.po) to git? My guess is few people running lookit-api locally will be touching these (esp. since people run it locally just to develop new frames!), so providing the compiled files would cut one step out of setup (although it might be an optional step anyway). If we don't check them in, we'll also need to add generating them (django-admin compilemessages) to the deployment pipeline.

4) Would you want to go ahead and merge this asap or wait on having the flat pages done?

**Comment by rhodricusack** [comment]
Thanks for the review!


0. The easiest way is to hit localhost:8000/ja/
You'll need to use the .po file I pushed through, as if you create one using makemessages it won't contain any Japanese content, and so will just render as English. The one I committed used "poedit" to guess translations (the real translations are under construction). I then took out tags that say #! fuzzy or similar, as django does not render these.

  1.  I have focussed on the participant facing text. We have not yet decided on the priority of translating experimenter-facing text.
  2.  I have implemented this but can't currently test it as I don't have any studies to test it on. Could one of you check it?
  3.  -
  4.  My goal here was to check out the pipeline, which seems to be working, so I don't mind if you merge or not. Apart from trying the i18n flatpages, our changes are likely to be mainly focussed on translation for a while, with only small changes to the code. So, it might make sense to merge, in case you have significant changes in the pipeline?

________________________________

**Comment by kimberscott** [comment]
> The easiest way is to hit localhost:8000/ja/

Ah-hah, thanks very much! A few more changes pushed to the review branch upon being able to see what was untranslated.

There are a few things that aren't behaving as I'd expect - does this make sense to you? E.g. the "Birthday" field in the child update form (defined in accounts/forms/ChildUpdateForm) isn't translated, but it's in the translation file and gets translated elsewhere, and the "Gender" field does get translated (but only if we use `_("Gender")` in the form labels. Also, the various options for some of the child questions aren't translated - see screenshot below:

Screen Shot 2020-08-19 at 2 08 20 PM

> I have implemented this but can't currently test it as I don't have any studies to test it on. Could one of you check it?

Sure - push changes up here for me to check? You can also merge in the changes at https://github.com/rhodricusack/lookit-api/pull/1 if those look ok to you (you'll need to add the additional translations though).

You can also create a study (Experimenter -> New study) if you make your account a researcher account, which you can do from the admin app (localhost:8000/__CTRL__). The protocol can be empty and you don't need to actually build the experiment runner. To make the study show up in the list, just find the study in Admin, check the "Built" and "Public" boxes, and set the State to "active".

One other question - what do you propose as the workflow for keeping translations up to date as any participant-facing text is added or edited? We can make sure to mark all text that goes to participants for translation. What do we do to update the translation files and note any new strings that need translations?

**Comment by rhodricusack** [comment]
@kimberscott sorry for the delay. I addressed the issue with the Gender, weeks and so on. The problem was that in fields.py we were using gettext not gettext_lazy, which means the translation was getting done when django starts its server, and not when the particular page is requested. I also fixed some missing translation fields around this.

I merged your changes from https://github.com/lookit/lookit-api/pull/527.

For some reason, I can't run the admin app. I can log in as an experimenter, and see "Manage studies" and "View Participants" and so on. But if I hit localhost:8000/__CTRL__ it takes me into an infinite loop of logging in (URL becomes localhost:8000/__CTRL__/login/?next=/__CTRL__) despite the fact it knows I'm logged in. Any idea what's going on there? 

On a pipeline for future changes, we propose the following. 
1. Some languages will be maintained, others will transition to "basis for future studies but requires work" status.
2. Following any changes to the codebase, you run django-admin makemessages to update language-specific .po files (presumably, in develop branch)
3. For each .po file in a maintained language, set up a watch that emails changes to the primary maintainer for that language.
4. That maintainer will make translations and push back to our fork. 
5. This will be reviewed by us and then a pull request submitted to you.

Ideally, this whole cycle from 2-5 would be executed before you commit changes to the production (master) branch so that the language translations are always up-to-date with the software. We will try to manage this effectively, but we are dependent on globally distributed volunteers so we will have to monitor how this is going.

Very best

**Comment by kimberscott** [comment]
> @kimberscott sorry for the delay. I addressed the issue with the Gender, weeks and so on. The problem was that in fields.py we were using gettext not gettext_lazy, which means the translation was getting done when django starts its server, and not when the particular page is requested. I also fixed some missing translation fields around this.

ah-hah, thanks for figuring that out!

> For some reason, I can't run the admin app. I can log in as an experimenter, and see "Manage studies" and "View Participants" and so on. But if I hit localhost:8000/CTRL it takes me into an infinite loop of logging in (URL becomes localhost:8000/CTRL/login/?next=/CTRL) despite the fact it knows I'm logged in. Any idea what's going on there?

Ack, that's a Markdown issue in my comment! You want `https://localhost:8000/__CTRL__` (with the double underscores on each side) - does that work?

Re: pipeline - @Datamance what do you think? We may want something that's not quite "every time anything goes to prod" (e.g., occasional quick bugfixes!) and it would be nice to add making the .po files to a precommit hook or something so they just stay up-to-date.

**Comment by rhodricusack** [comment]
I knew about the double underscores, I could see it in the code. You'll have to interpret my infinite loop response through the lens of the same Markdown issue. So that isn't it. Should I be using a regular "Experimenter" password? This is what I've been trying. Should it instead be the Django superuser or something? What's the URL of a page it goes to after it has authenticated? Perhaps I can just skip to that. Do you have any other ideas?



Rhodri Cusack
Thomas Mitchell Professor of Cognitive Neuroscience
Trinity College Institute of Neuroscience
School of Psychology
Trinity College Dublin
Dublin 2
Ireland

Lloyd Institute, Room 3.21
www.cusacklab.org<http://www.cusacklab.org>
+353 1 896 2977 (w)



________________________________

**Comment by Datamance** [comment]
> Should it instead be the Django superuser or something?

In the [local setup instructions](https://lookit.readthedocs.io/en/develop/install-django-project.html#basic-installation) there's a note to create a superuser with `python manage.py createsuperuser` make sure you do that first.

**Review by Datamance — CHANGES_REQUESTED** [review]
Just a small nitpick - make sure there's a newline at EOF in the settings file

File: project/settings/defaults.py
Comment (rhodricusack) [reply project/settings/defaults.py]: I've done this!

**Comment by rhodricusack** [comment]
Is there any chance we could complete this pull request? I would like to shift the repo that I'm using from my own account to a manybabies github account, before giving the translators access, and I think the simplest way to achieve that would be to merge in these changes and then delete this repo?

Hope all is well!

**Review by Datamance — APPROVED** [review]

**Comment by Datamance** [comment]
@rhodricusack I'm going to approve and merge this - let it be noted here, however, that this is technically going outside of the [soon-to-be-merged documentation](https://github.com/lookit/lookit-api/pull/579/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R150) around the localization process in the sense that your translators have already added the `.mo` files for their locales (originally, I was intending to add a `compilemessages` job to the deployment pipeline to do this at the last possible minute). 

Originally, I wasn't super hot on the idea of adding compiled binaries to source control; on the other hand, I want to QA this stuff in staging and there's some nontrivial setup with Kubernetes (health checks etc.) to get the jobs running in order. Let's just merge this in for now so we can QA. Depending on how it goes, I may change the documented process to just permanently do the message compilation upstream!

**Comment by rhodricusack** [comment]
@Datamance Thank you. Let me know when you have had a chance to set up the automatic compilation and I will remove the .mo files.

**Pull request merged**
